Ports Abductors from Starlight

gluesniffler opened the pull request:
## About the PR
ayy lmao

This PR ports abductors from [Starlight](https://github.com/ss14Starlight/space-station-14/) [(Discord)](https://discord.gg/wAyQKB78fH) with a few more additions to make it extra goob-y - The old copypasta license joke is overdone so I won't do it :)

## Technical Details

Patch is CC0-1.0 (except for sprites and other stuff taken from 13 ofc). Feel free to yoink :blunt:

## Why / Balance
content content content content content content (also i was paid 4k goobcoin for this)

## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.
- [X] I am ready 4 ze new tider order

## Credits
- Again, props to [Starlight](https://discord.gg/wAyQKB78fH) for all of the work on these!
- Big ups to Darkrell, Rinary and Landosaur for making the funny ayyliens real.

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
:cl: Mocho, Darkrell, Rinary1, Landosaur
- add: Voople xop narp gleeple xop verp narple xip "abductor sightings" gleeple narple xip voop.
- add: Voople narp narp "superior organ" gleeple narple "side effects" glimp xop voople narp.
- add: Voople glimple xop "mental crashout" verle gleep narple xoop vop "feral tiders" glip verple xop narp.
- tweak: Advanced surgical tools now need to be toggled between one of two modes at a time. The medical omnitool remains unchanged.

gluesniffler commented:
Okay so off the top of my head I made two war crimes:

1. I dont fucking know how to properly use GameRules if I want to assign antag status to ONLY one specific person after an event, so I used ForceMakeAntag on some of these.
2. Not sure how weights work exactly and to calculate their odds, generally I'm looking for 70/25/5ish kinds of splits.

Besides that, no clue what else might be iffy.

whateverusername0 commented on Content.Client/NetworkConfigurator/Systems/NetworkConfiguratorSystem.cs:
```diff
@@ -114,7 +114,8 @@ public StatusControl(NetworkConfiguratorComponent configurator, string keyBindin
             _keyBindingName = keyBindingName;
             _label = new RichTextLabel { StyleClasses = { StyleNano.StyleClassItemStatus } };
-            AddChild(_label);
+            if (_configurator.ShowLabel)
```
huh?

gluesniffler replied:
Oh, forgot to comment it, its used to hide the Link/List mode on the abductor's multitool, which has an alt mode now that serves like a doorjack in 13

JustSome1Nearby commented:
I think there is a BIG legal problem with Project Starlight licenses and stuff

JustSome1Nearby commented:
But ... If it works... Who cares?

gluesniffler commented:
> I think there is a BIG legal problem with Project Starlight licence and stuff

they have a new license, tldr is "MIT but pls credit"

JustSome1Nearby commented:
By the way... How done is it?

(IF it is ready enough not to crash the server, "speedmerge now")

gluesniffler commented on Content.Shared/_Shitmed/ItemSwitch/SharedItemSwitchSystem.cs:
```diff
@@ -0,0 +1,186 @@
+    [Dependency] private readonly SharedItemSystem _item = default!;
+    [Dependency] private readonly ClothingSystem _clothing = default!;
+
+    private EntityQuery<ItemSwitchComponent> _query;
```
It seems to be working fine at least on all cases where im using it 🤷

gluesniffler commented on Content.Server/_Shitmed/Antags/Abductor/AbductorSystem.Console.cs:
```diff
@@ -0,0 +1,159 @@
+    private void OnAbductGetProgress(Entity<AbductConditionComponent> ent, ref ObjectiveGetProgressEvent args)
+        => args.Progress = DoorjackProgress(ent.Comp, _number.GetTarget(ent.Owner));
+
+    private float DoorjackProgress(AbductConditionComponent comp, int target)
```
i think they just copy and pasted this from ninja code and forgot to rename it lol

gluesniffler commented on Content.Shared/_Shitmed/ItemSwitch/Components/ItemSwitchComponent.cs:
```diff
@@ -0,0 +1,89 @@
+    /// /// <remarks>
+    /// If server-side systems affect the item's toggle, like charge/fuel systems, then the item is not predictable.
+    /// </remarks>
+    [ViewVariables(VVAccess.ReadWrite), DataField, AutoNetworkedField]
```
good question, now that i notice it, not sure why they also have the VV in here when theres a datafield, my guess is another copy paste job

gluesniffler commented on Content.Server/_Shitmed/Antags/Abductor/AbductorSystem.Actions.cs:
```diff
@@ -0,0 +1,145 @@
+    {
+        EnsureComp<AbductorScientistComponent>(ev.Performer, out var comp);
+
+        _color.RaiseEffect(Color.FromHex("#BA0099"), new List<EntityUid>(1) { ev.Performer }, Filter.Pvs(ev.Performer, entityManager: EntityManager));
```
Code is slightly different in the methods that use this, for now I'll just abstract the really large one for TeleportationEffectEntity(s)

Aviu00 reviewed (changes requested):
Tested it in game. For some reason camera console wasn't working for me, it was just static noise.
Abductors can't bump open doors and don't have footstep sounds (missing tags).
Guidebook entry exists apparently, but I didn't find a way to access it. Should be under the "antagonists" section and also question mark button leading to the entry on all abductor equipment.
Abductor equipment should be marked as contra.
If a player attaches to an abductor entity (via disconnecting and reconnecting, or "control mob" admin context menu), the server crashes.
It would be nice if the wonderprod showed currently selected mode in the description.

gluesniffler commented:
Fixed camera error, crash issue and guidebook on local

Currently working on figuring out the fact that the spawned eye seems to inherit the rotation of the abductor's grid, so its **never** flush with the station. And if you were to change directions during this, then the controls get all fucky...

gluesniffler commented:
All errors should be fixed, changelog is:

- I had missed some changes that Starlight made in SharedMoverController to allow entity movement to sync up across grid with relaying.
- Added state labels to items that use ItemSwitchComponent, those being the adv medical tools, and the abductor's wonderprod.
- Linked properly the guidebooks (forgot to create the entry in the goob file whoops)
- Added the funny tags for footsteps, let me know if I missed any other since this is basically a new species.
- Fixed the crash bug which was caused by Starlight code (?) apparently not properly setting the Eye's target to null on the server side, instead being set to 0, which would crash the server when trying to go back to the entity.
- Marked the combat-inclined gear as RestrictedContraband and MedicalScienceContraband for the more useful tools.
- Camera console was a skill issue from my end since I accidentally removed some lines of code from AI system lol

Aviu00 commented on Content.Shared/Emag/Systems/EmagSystem.cs:
```diff
@@ -56,6 +58,20 @@ public bool TryUseEmag(EntityUid uid, EntityUid user, EntityUid target, EmagComp
             return false;
         }
 
+        // Shitmed - Starlight Abductors: Check if the target has a whitelist, and check if it passes
+        if (!_whitelist.IsWhitelistFail(comp.ValidTargets, target))
+        {
+            _popup.PopupClient(Loc.GetString("emag-attempt-failed", ("tool", uid)), user, user);
+            return false;
+        }
+
+        // Shitmed - Starlight Abductors: Check if the target has a whitelist, and check if it passes
+        if (!_whitelist.IsWhitelistFail(comp.ValidTargets, target))
+        {
+            _popup.PopupClient(Loc.GetString("emag-attempt-failed", ("tool", uid)), user, user);
+            return false;
+        }
```
Why there are 2 exact same if statements? Also, I think there it shouldn't have '!', so abductor multitool works on airlocks and not on anything else except airlocks.

Aviu00 reviewed (changes requested):
Abductor multitool should be marked as contra since it's emag. Also, guidebook should actually tell that multitool can emag things.
And "send yourself" action on abductor console doesn't work for me.
Abductors still can't bump open doors.

gluesniffler commented:
> Abductor multitool should be marked as contra since it's emag. Also, guidebook should actually tell that multitool can emag things. And "send yourself" action on abductor console doesn't work for me. Abductors still can't bump open doors.

- Send yourself only works on the same grid (you spawn on another so you need to FTL), and within a certain distance of the station.
- Turns out the fucking abductor spawner redeclares the tags component AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA I HATE TAGS
- Emag stuff I probably copy pasted by accident. Fixed the wonky check, turns out I had inverted the logic because I forgot to set validTargets to AutoNetworkedField, which would lead to the client saying it was emagged anyway :clueless:
- Marked the stuff as contra and added guidebook entry

Aviu00 commented:
>  Send yourself only works on the same grid (you spawn on another so you need to FTL), and within a certain distance of the station.

At least something should tell you that.

gluesniffler commented:
> > Send yourself only works on the same grid (you spawn on another so you need to FTL), and within a certain distance of the station.
> 
> At least something should tell you that.

idk how to do that lol, putting it in the guidebook probs wont be effective enough, and i dont think theres anything in WorldTargetAction events that lets you do that

Piras314 commented:
Seems silly that this is using the _Shitmed namespace but it's not a big deal

Aviu00 commented:
> idk how to do that lol, putting it in the guidebook probs wont be effective enough, and i dont think theres anything in WorldTargetAction events that lets you do that

Maybe it's better to make it instant action that teleports you to eye position instead?

gluesniffler commented:
> Seems silly that this is using the _Shitmed namespace but it's not a big deal

the antag literally doesnt work without surgery so it was more fitting to put it there

gluesniffler commented:
> > idk how to do that lol, putting it in the guidebook probs wont be effective enough, and i dont think theres anything in WorldTargetAction events that lets you do that
> 
> Maybe it's better to make it instant action that teleports you to eye position instead?

I feel the targeting makes more sense as it prevents you from accidentally pressing the button and now you get sent on station, instead you get an implicit confirmation. 

Don't think I have enough patience left after dealing with the file hunting to go figure out how to do cross-grid warnings on this thing either.

Aviu00 commented:
> I feel the targeting makes more sense as it prevents you from accidentally pressing the button and now you get sent on station, instead you get an implicit confirmation.

Simply make it confirmable action like suicide actions work then.

caoimhan2008 commented:
based

caoimhan2008 commented:
bruh

Piras314 reviewed (commented):
LGTM but I will test this before merging.

Aviu00 commented:
Isn't a single abductor that is not even allowed to kill and do stuff like that is kinda unsuitable for roundstart antag? You can make it spawn similarly to heretics and thieves if you want it to be roundstart. And if it is roundstart, it should be selectable in antagonists menu and roll before jobs.
Also make it work with dynamic.

gluesniffler commented:
> Isn't a single abductor that is not even allowed to kill and do stuff like that is kinda unsuitable for roundstart antag? You can make it spawn similarly to heretics and thieves if you want it to be roundstart. And if it is roundstart, it should be selectable in antagonists menu and roll before jobs. Also make it work with dynamic.

i aint touching dynamic shitcode with a 10ft pole

and i added it to the roundstart stuff at piras's request, they can modify it if they want

JustSome1Nearby commented:
> Isn't a single abductor that is not even allowed to kill and do stuff like that is kinda unsuitable for roundstart antag? You can make it spawn similarly to heretics and thieves if you want it to be roundstart. And if it is roundstart, it should be selectable in antagonists menu and roll before jobs. Also make it work with dynamic.

Also, to be fair, abductor feels like the best suited antag to actualy have fun playing, just mad surgeon simulator with absolutely no consequences whatssoever! (Proceeds to steal the Captain. Shuttle gets found with a pinpointer. Joeover.png)

caoimhan2008 commented:
> Isn't a single abductor that is not even allowed to kill and do stuff like that is kinda unsuitable for roundstart antag? You can make it spawn similarly to heretics and thieves if you want it to be roundstart. And if it is roundstart, it should be selectable in antagonists menu and roll before jobs. Also make it work with dynamic.

abductor can kill if it has to (eg to survive)

caoimhan2008 commented:
your point also doesnt make any sense as thief is a roundstart antag that literally cant kill

Piras314 reviewed (changes requested):
aviu's latest comment is right (though I don't think you have to add it to dynamic in the main pr, it's easy to do so iirc but dynamic doesn't work so it's not a big deal), make it selectable in the antags menu and make it roll before jobs

caoimhan2008 commented:
> aviu's latest comment is right (though I don't think you have to add it to dynamic in the main pr, it's easy to do so iirc but dynamic doesn't work so it's not a big deal), make it selectable in the antags menu and make it roll before jobs

pretty sure he already did that

caoimhan2008 commented:
merge when

deltanedas commented:
when you abduct it

starch70 commented:
did oyu fucking readd tidemind bro

JustSome1Nearby commented:
> did oyu fucking readd tidemind bro

Re add?! What?!

(Also, invincible cable cuffs from wounderprod gonna be fixed?)

JustSome1Nearby commented:
So looking at how it plays ...

Wounderprod can be used by people (I am reading on the ss13 wiki that it isnt supposed to? It flies out of your hands with any slip or push, swings are slooow… so isnt really a GREAT weapon)
Cuffs still are unbreakable?

Your Suit, Helmet, Toolbelt, are all removable (suit is supposed to be surgicaly attached, and other parts just make sense to not be removed by tiders... Maybe add there cluwn''s unremovable clothing component? Else abductors are just looted)

There are no Chameleon Vest interactions, no way to blend in with the crowd (Maybe since you are a solo abductor, not sure)
(Also cause your clothes can be removed , lrp Abductors may just start changing them, wich isnt supposed to work like that?)

There is no Silencer or Decloner

Abductor ship only has brute and burn meds, nothing against poison and stuff

JustSome1Nearby commented:
Also, all glands seem to be unnamed, its not like crew will steal them, just let me see what fate am i subjecting them to!

gluesniffler commented:
> So looking at how it plays ...
> 
> Wounderprod can be used by people (I am reading on the ss13 wiki that it isnt supposed to? It flies out of your hands with any slip or push, swings are slooow… so isnt really a GREAT weapon) Cuffs still are unbreakable?
> 
> Your Suit, Helmet, Toolbelt, are all removable (suit is supposed to be surgicaly attached, and other parts just make sense to not be removed by tiders... Maybe add there cluwn''s unremovable clothing component? Else abductors are just looted)
> 
> There are no Chameleon Vest interactions, no way to blend in with the crowd (Maybe since you are a solo abductor, not sure) (Also cause your clothes can be removed , lrp Abductors may just start changing them, wich isnt supposed to work like that?)
> 
> There is no Silencer or Decloner
> 
> Abductor ship only has brute and burn meds, nothing against poison and stuff

- I left wonderprod usable by tiders because its very funny, and antag gear being good incentivizes validhunting the abductor rather than letting them run free :trollface:

- If you get looted its squarely your fault considering that you have a literal free use teleport back. Much like a nukie getting stripped of their gear by tiders and turning tail in shame.

- Lone abductor isnt supposed to be stealthy either, you go in hot, do your thing, fuck off. Duo abductor is another thing entirely.

- WYPI

- There's no topical medication against poison and stuff, and the only reason you'd have to heal your experimental subject is if they're dead, at which point it doesn't really matter if you have it or not.

- Glands will remain as is, learn what each icon commonly is and try to take a gander :godo:
